File: packages/sdk-python/examples/support_agent.py
"""
Tollgate Example: AI Support Agent

A support agent backed by Claude that uses Tollgate to gate risky actions.
Three tools are protected: issue_refund, update_account, escalate_to_human.

Run: python support_agent.py
"""

import os
import sys

import anthropic

from tollgate import ActionDenied, ActionPending, Tollgate

TOLLGATE_API_KEY = os.environ.get("TOLLGATE_API_KEY", "")
TOLLGATE_BASE_URL = os.environ.get("TOLLGATE_BASE_URL", "http://localhost:8000")
ANTHROPIC_API_KEY = os.environ.get("ANTHROPIC_API_KEY", "")

if not TOLLGATE_API_KEY:
    print("Error: TOLLGATE_API_KEY is not set.")
    sys.exit(1)
if not ANTHROPIC_API_KEY:
    print("Error: ANTHROPIC_API_KEY is not set.")
    sys.exit(1)

tg = Tollgate(
    api_key=TOLLGATE_API_KEY,
    base_url=TOLLGATE_BASE_URL,
    on_pending=lambda action_id: print(f"\n[Tollgate] Waiting for approval in Slack... (action_id={action_id})"),
    poll_interval=3.0,
    max_wait=300.0,
)

client = anthropic.Anthropic(api_key=ANTHROPIC_API_KEY)

TOOLS = [
    {
        "name": "issue_refund",
        "description": "Issue a refund to a customer for a specified amount.",
        "input_schema": {
            "type": "object",
            "properties": {
                "amount": {"type": "number", "description": "Refund amount in USD"},
                "customer_id": {"type": "string", "description": "Customer identifier"},
                "reason": {"type": "string", "description": "Reason for the refund"},
            },
            "required": ["amount", "customer_id"],
        },
    },
    {
        "name": "update_account",
        "description": "Update a customer's account information.",
        "input_schema": {
            "type": "object",
            "properties": {
                "customer_id": {"type": "string", "description": "Customer identifier"},
                "field": {"type": "string", "description": "Field to update (email, plan, etc.)"},
                "value": {"type": "string", "description": "New value for the field"},
            },
            "required": ["customer_id", "field", "value"],
        },
    },
    {
        "name": "escalate_to_human",
        "description": "Escalate this support case to a human agent.",
        "input_schema": {
            "type": "object",
            "properties": {
                "customer_id": {"type": "string", "description": "Customer identifier"},
                "reason": {"type": "string", "description": "Reason for escalation"},
                "priority": {"type": "string", "enum": ["low", "medium", "high"], "description": "Priority level"},
            },
            "required": ["customer_id", "reason"],
        },
    },
]


@tg.guard("issue_refund")
def issue_refund(amount: float, customer_id: str, reason: str = "") -> dict:  # type: ignore[return]
    """Execute the refund (would call Stripe in production)."""
    print(f"[Tool] Issuing refund: ${amount} to {customer_id}")
    return {"status": "refunded", "amount": amount, "customer_id": customer_id}


@tg.guard("update_account")
def update_account(customer_id: str, field: str, value: str) -> dict:  # type: ignore[return]
    """Execute the account update (would call your DB in production)."""
    print(f"[Tool] Updating {field}={value} for {customer_id}")
    return {"status": "updated", "customer_id": customer_id, "field": field, "value": value}


@tg.guard("escalate_to_human")
def escalate_to_human(customer_id: str, reason: str, priority: str = "medium") -> dict:  # type: ignore[return]
    """Escalate to a human agent."""
    print(f"[Tool] Escalating {customer_id} — {reason} (priority={priority})")
    return {"status": "escalated", "customer_id": customer_id, "ticket_id": "TKT-001"}


def run_tool(name: str, inputs: dict) -> str:  # type: ignore[type-arg]
    """Dispatch a tool call, catching Tollgate exceptions."""
    try:
        if name == "issue_refund":
            result = issue_refund(**inputs)
        elif name == "update_account":
            result = update_account(**inputs)
        elif name == "escalate_to_human":
            result = escalate_to_human(**inputs)
        else:
            return f"Unknown tool: {name}"
        return str(result)
    except ActionDenied as e:
        return f"[Tollgate] Action denied: {e.reason}"
    except ActionPending as e:
        return f"[Tollgate] Approval timed out after {e.timeout_seconds}s for action {e.action_id}"


def chat_loop() -> None:
    """Run the support agent loop."""
    print("Tollgate Support Agent — type 'quit' to exit")
    print(f"Connected to Tollgate at {TOLLGATE_BASE_URL}\n")
    messages: list[dict] = []  # type: ignore[type-arg]

    while True:
        try:
            user_input = input("You: ").strip()
        except (EOFError, KeyboardInterrupt):
            print("\nGoodbye.")
            break

        if user_input.lower() in ("quit", "exit", "q"):
            print("Goodbye.")
            break
        if not user_input:
            continue

        messages.append({"role": "user", "content": user_input})

        response = client.messages.create(
            model="claude-haiku-4-5-20251001",
            max_tokens=1024,
            system=(
                "You are a helpful customer support agent. "
                "Use the available tools to help customers with refunds, account updates, and escalations. "
                "Always confirm the customer_id before taking actions."
            ),
            tools=TOOLS,  # type: ignore[arg-type]
            messages=messages,  # type: ignore[arg-type]
        )

        while response.stop_reason == "tool_use":
            tool_results = []
            assistant_content = []

            for block in response.content:
                assistant_content.append(block)
                if block.type == "tool_use":
                    print(f"\n[Agent] Calling {block.name}({block.input})")
                    result = run_tool(block.name, block.input)  # type: ignore[arg-type]
                    tool_results.append({
                        "type": "tool_result",
                        "tool_use_id": block.id,
                        "content": result,
                    })

            messages.append({"role": "assistant", "content": assistant_content})
            messages.append({"role": "user", "content": tool_results})

            response = client.messages.create(
                model="claude-haiku-4-5-20251001",
                max_tokens=1024,
                system=(
                    "You are a helpful customer support agent. "
                    "Use the available tools to help customers with refunds, account updates, and escalations."
                ),
                tools=TOOLS,  # type: ignore[arg-type]
                messages=messages,  # type: ignore[arg-type]
            )

        # Extract final text response
        reply = ""
        for block in response.content:
            if hasattr(block, "text"):
                reply += block.text

        messages.append({"role": "assistant", "content": response.content})
        print(f"\nAgent: {reply}\n")


if __name__ == "__main__":
    chat_loop()

File: packages/sdk-typescript/examples/support-agent.ts
/**
 * Support Agent Example
 *
 * Demonstrates a Claude-powered support agent with three guarded tools,
 * each gated by the Tollgate policy layer before execution.
 *
 * Run:
 *   TOLLGATE_API_KEY=sk-... ANTHROPIC_API_KEY=sk-... npx tsx examples/support-agent.ts
 */

import Anthropic from "@anthropic-ai/sdk";
import { Tollgate } from "@tollgate/sdk";

// ---------------------------------------------------------------------------
// Clients
// ---------------------------------------------------------------------------

const anthropic = new Anthropic({
  apiKey: process.env.ANTHROPIC_API_KEY,
});

const tg = new Tollgate({
  apiKey: process.env.TOLLGATE_API_KEY!,
  baseUrl: process.env.TOLLGATE_BASE_URL ?? "http://localhost:8000",
  failOpen: false,
  pollIntervalMs: 1000,
  pollTimeoutMs: 60_000,
});

// ---------------------------------------------------------------------------
// Guarded tool implementations
// ---------------------------------------------------------------------------

const issueRefund = tg.guard(
  "issue_refund",
  async (orderId: string, amount: number): Promise<string> => {
    // In production: call your payment provider API here
    console.log(`[tool] Issuing refund for order ${orderId}: $${amount}`);
    return `Refund of $${amount} issued for order ${orderId}. Confirmation #REF-${Date.now()}.`;
  },
  (orderId, amount) => ({ order_id: orderId, amount }),
);

const deleteAccount = tg.guard(
  "delete_account",
  async (customerId: string): Promise<string> => {
    // In production: call your account management API here
    console.log(`[tool] Deleting account for customer ${customerId}`);
    return `Account ${customerId} has been scheduled for deletion. Data will be purged in 30 days.`;
  },
  (customerId) => ({ customer_id: customerId }),
);

const escalateToHuman = tg.guard(
  "escalate_to_human",
  async (ticketId: string, reason: string): Promise<string> => {
    // In production: create a ticket in your support system
    console.log(`[tool] Escalating ticket ${ticketId}: ${reason}`);
    return `Ticket ${ticketId} escalated to human support team. Reason: ${reason}. ETA: 2-4 hours.`;
  },
  (ticketId, reason) => ({ ticket_id: ticketId, reason }),
);

// ---------------------------------------------------------------------------
// Claude tool definitions
// ---------------------------------------------------------------------------

const tools: Anthropic.Tool[] = [
  {
    name: "issue_refund",
    description:
      "Issue a monetary refund to a customer for a given order. Requires policy approval.",
    input_schema: {
      type: "object" as const,
      properties: {
        order_id: {
          type: "string",
          description: "The order ID to refund",
        },
        amount: {
          type: "number",
          description: "The refund amount in USD",
        },
      },
      required: ["order_id", "amount"],
    },
  },
  {
    name: "delete_account",
    description:
      "Permanently delete a customer account and all associated data. Requires policy approval.",
    input_schema: {
      type: "object" as const,
      properties: {
        customer_id: {
          type: "string",
          description: "The customer ID to delete",
        },
      },
      required: ["customer_id"],
    },
  },
  {
    name: "escalate_to_human",
    description:
      "Escalate a support ticket to a human agent when the issue cannot be resolved automatically.",
    input_schema: {
      type: "object" as const,
      properties: {
        ticket_id: {
          type: "string",
          description: "The ticket ID to escalate",
        },
        reason: {
          type: "string",
          description: "The reason for escalation",
        },
      },
      required: ["ticket_id", "reason"],
    },
  },
];

// ---------------------------------------------------------------------------
// Tool dispatcher
// ---------------------------------------------------------------------------

async function dispatchTool(
  toolName: string,
  toolInput: Record<string, unknown>,
): Promise<string> {
  try {
    switch (toolName) {
      case "issue_refund":
        return await issueRefund(
          toolInput.order_id as string,
          toolInput.amount as number,
        );

      case "delete_account":
        return await deleteAccount(toolInput.customer_id as string);

      case "escalate_to_human":
        return await escalateToHuman(
          toolInput.ticket_id as string,
          toolInput.reason as string,
        );

      default:
        return `Unknown tool: ${toolName}`;
    }
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    return `Tool execution failed: ${message}`;
  }
}

// ---------------------------------------------------------------------------
// Support agent loop
// ---------------------------------------------------------------------------

async function runSupportAgent(userMessage: string): Promise<void> {
  console.log(`\nUser: ${userMessage}\n`);

  const messages: Anthropic.MessageParam[] = [
    { role: "user", content: userMessage },
  ];

  let continueLoop = true;

  while (continueLoop) {
    const response = await anthropic.messages.create({
      model: "claude-opus-4-5",
      max_tokens: 1024,
      system:
        "You are a helpful customer support agent. You have access to tools to issue refunds, delete accounts, and escalate tickets. " +
        "Use tools when appropriate to help customers. Be concise and professional.",
      tools,
      messages,
    });

    // Collect text output
    for (const block of response.content) {
      if (block.type === "text") {
        console.log(`Assistant: ${block.text}`);
      }
    }

    if (response.stop_reason === "end_turn") {
      continueLoop = false;
      break;
    }

    if (response.stop_reason === "tool_use") {
      // Process tool calls
      const toolResults: Anthropic.ToolResultBlockParam[] = [];

      for (const block of response.content) {
        if (block.type === "tool_use") {
          console.log(`\n[Calling tool: ${block.name}]`);
          const result = await dispatchTool(
            block.name,
            block.input as Record<string, unknown>,
          );
          console.log(`[Tool result: ${result}]\n`);

          toolResults.push({
            type: "tool_result",
            tool_use_id: block.id,
            content: result,
          });
        }
      }

      // Append assistant turn and tool results
      messages.push({ role: "assistant", content: response.content });
      messages.push({ role: "user", content: toolResults });
    } else {
      continueLoop = false;
    }
  }
}

// ---------------------------------------------------------------------------
// Entry point
// ---------------------------------------------------------------------------

const scenario = process.argv[2] ?? "refund";

const scenarios: Record<string, string> = {
  refund:
    "Hi, I ordered product #ORDER-789 last week and it arrived damaged. I'd like a full refund of $59.99 please.",
  delete:
    "I want to close my account permanently. My customer ID is CUST-12345. Please delete all my data.",
  escalate:
    "I've been trying to resolve ticket TKT-5678 for two weeks with no success. I need to speak to a manager.",
};

const message = scenarios[scenario] ?? scenarios.refund;

runSupportAgent(message).catch((err) => {
  console.error("Agent error:", err);
  process.exit(1);
});
